packages/inventory-next/src/services/inventory.ts

```diff
@@ -0,0 +1,1034 @@
+import { InternalModuleDeclaration } from "@medusajs/modules-sdk"
+import {
+  Context,
+  IInventoryServiceNext,
+  InventoryTypes,
+  ModuleJoinerConfig,
+  ModulesSdkTypes,
+  InventoryNext,
+  ReservationItemDTO,
+} from "@medusajs/types"
+import {
+  EmitEvents,
+  MedusaContext,
+  MedusaError,
+  ModulesSdkUtils,
+} from "@medusajs/utils"
+import { entityNameToLinkableKeysMap, joinerConfig } from "../joiner-config"
+import { InventoryItem, InventoryLevel, ReservationItem } from "@models"
+import { DAL } from "@medusajs/types"
+import { InjectTransactionManager } from "@medusajs/utils"
+import { InjectManager } from "@medusajs/utils"
+import InventoryLevelService from "./inventory-level"
+import { partitionArray } from "@medusajs/utils"
+import { InventoryEvents } from "@medusajs/utils"
+import { CommonEvents } from "@medusajs/utils"
+import { isDefined } from "@medusajs/utils"
+
+type InjectedDependencies = {
+  baseRepository: DAL.RepositoryService
+  inventoryItemService: ModulesSdkTypes.InternalModuleService<any>
+  inventoryLevelService: InventoryLevelService<any>
+  reservationItemService: ModulesSdkTypes.InternalModuleService<any>
+}
+
+const generateMethodForModels = [InventoryItem, InventoryLevel, ReservationItem]
+
+export default class InventoryModuleService<
+    TInventoryItem extends InventoryItem = InventoryItem,
+    TInventoryLevel extends InventoryLevel = InventoryLevel,
+    TReservationItem extends ReservationItem = ReservationItem
+  >
+  extends ModulesSdkUtils.abstractModuleServiceFactory<
+    InjectedDependencies,
+    InventoryNext.InventoryItemDTO,
+    {
+      InventoryItem: {
+        dto: InventoryNext.InventoryItemDTO
+      }
+      InventoryLevel: {
+        dto: InventoryNext.InventoryLevelDTO
+      }
+      ReservationItem: {
+        dto: InventoryNext.ReservationItemDTO
+      }
+    }
+  >(InventoryItem, generateMethodForModels, entityNameToLinkableKeysMap)
+  implements IInventoryServiceNext
+{
+  protected baseRepository_: DAL.RepositoryService
+
+  protected readonly inventoryItemService_: ModulesSdkTypes.InternalModuleService<TInventoryItem>
+  protected readonly reservationItemService_: ModulesSdkTypes.InternalModuleService<TReservationItem>
+  protected readonly inventoryLevelService_: InventoryLevelService<TInventoryLevel>
+
+  constructor(
+    {
+      baseRepository,
+      inventoryItemService,
+      inventoryLevelService,
+      reservationItemService,
+    }: InjectedDependencies,
+    protected readonly moduleDeclaration?: InternalModuleDeclaration
+  ) {
+    // @ts-ignore
+    // eslint-disable-next-line prefer-rest-params
+    super(...arguments)
+
+    this.baseRepository_ = baseRepository
+    this.inventoryItemService_ = inventoryItemService
+    this.inventoryLevelService_ = inventoryLevelService
+    this.reservationItemService_ = reservationItemService
+  }
+
+  __joinerConfig(): ModuleJoinerConfig {
+    return joinerConfig
+  }
+
+  private async ensureInventoryLevels(
```

Comment:
The naming is a bit confusing, shouldn't it just be `listInventoryLevels`?

Reply:
it's not exactly `listInventoryLevels` as the purpose is to validate that the inventory levels being requested for creating reservations/updating levels and the arguments are distinct from what would be expected for `listInventoryLevels` based on what we usually provide as arguments. It's also not in the expected behavior of our listEndpoints to throw if all records are not found based on the provided criteria.

 We could rename this as `ensureInventoryLevelsExist`, `listAndValidateInventoryLevels` or `listInventoryLevelsOrThrow` instead, but I don't this just calling this `listInventoryLevels` is an appropriate solution. wdyt?

Reply:
listAndValidateInventoryLevels sounds good to me, as we actually do both here 👍